# Project: BCG Company Dataset Investigation

## Introduction:
> In this project we are using BCG dataset which is an electricity and gas company and we are trying to predict whether the client is going to churn (stop doing business with BCG) or not. And so, we can make a model that provides clients who are expected to churn with a 20% discount on the product/service.

## Columns' Description
> In client.csv:
- **id** = client company identifier
- **channel_sales** = code of the sales channel
- **cons_12m** = electricity consumption of the past 12 months
- **cons_gas_12m** = gas consumption of the past 12 months
- **cons_last_month** = electricity consumption of the last month
- **date_activ** = date of activation of the contract
- **date_end** = date of the end of the contract
- **date_modif_prod** = date of the last modification of the product
- **date_renewal** = date of the next contract renewal
- **forecast_cons_12m** = expected electricity consumption for next 12 months
- **forecast_cons_year** = expected electricity consumption for the next calendar year
- **forecast_discount_energy** = expected value of current discount
- **forecast_meter_rent_12m** = expected bill of meter rental for the next 12 months
- **forecast_price_energy_off_peak** = expected energy price for 1st period (off peak)
- **forecast_price_energy_peak** = expected energy price for 2nd period (peak)
- **forecast_price_pow_off_peak** = expected power price for 1st period (off peak)
- **has_gas** =  client is also a gas client
- **imp_cons** = current paid consumption
- **margin_gross_pow_ele** = gross margin on power subscription
- **margin_net_pow_ele** = net margin on power subscription
- **nb_prod_act** = number of active products and services
- **net_margin** = total net margin
- **num_years_antig** = how many years did the client stay with bcg.
- **origin_up** = code of the electricity campaign the customer first subscribed to
- **pow_max** = subscribed power
- **churn** = has the client churned over the next 3 months

> In price.csv
- **id** = client company identifier
- **price_date** = reference date
- **price_off_peak_var** = price of energy for the 1st period (off peak)
- **price_peak_var** = price of energy for the 2nd period (peak)
- **price_mid_peak_var** = price of energy for the 3rd period (mid peak)
- **price_off_peak_fix** = price of power for the 1st period (off peak)
- **price_peak_fix** = price of power for the 2nd period (peak)
- **price_mid_peak_fix** = price of power for the 3rd period (mid peak)

> **Note:** Some fields are hashed text strings. This preserves the privacy of the original data 



### Part I: Data Wrangling

In [1]:
#Importing Libraries.
from matplotlib import pyplot 
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

# Shows plots in jupyter notebook
%matplotlib inline

# Set plot style
sns.set(color_codes=True)
pd.set_option("max_columns", None)

In [2]:
#Reading client and prices data
client_df=pd.read_csv('client_data.csv')
client_df

id                     channel_sales  \
0      24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1      d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2      764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3      bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4      149d57cf92fc41cf94415803a877cb4b                           MISSING   
...                                 ...                               ...   
14601  18463073fb097fc0ac5d3e040f356987  foosdfpfkusacimwkcsosbicdxkicaua   
14602  d0a6f71671571ed83b2645d23af6de00  foosdfpfkusacimwkcsosbicdxkicaua   
14603  10e6828ddd62cbcf687cb74928c4c2d2  foosdfpfkusacimwkcsosbicdxkicaua   
14604  1cf20fd6206d7678d5bcafd28c53b4db  foosdfpfkusacimwkcsosbicdxkicaua   
14605  563dde550fd624d7352f3de77c0cdfcd                           MISSING   

       cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0             0         54946                0  2013-06-15  2016-06-15   
1          4660             0                0  2009-08-21  2016-08-30   
2           544             0                0  2010-04-16  2016-04-16   
3          1584             0                0  2010-03-30  2016-03-30   
4          4425             0              526  2010-01-13  2016-03-07   
...         ...           ...              ...         ...         ...   
14601     32270         47940                0  2012-05-24  2016-05-08   
14602      7223             0              181  2012-08-27  2016-08-27   
14603      1844             0              179  2012-02-08  2016-02-07   
14604       131             0                0  2012-08-30  2016-08-30   
14605      8730             0                0  2009-12-18  2016-12-17   

      date_modif_prod date_renewal  forecast_cons_12m  forecast_cons_year  \
0          2015-11-01   2015-06-23               0.00                   0   
1          2009-08-21   2015-08-31             189.95                   0   
2          2010-04-16   2015-04-17              47.96                   0   
3          2010-03-30   2015-03-31             240.04                   0   
4          2010-01-13   2015-03-09             445.75                 526   
...               ...          ...                ...                 ...   
14601      2015-05-08   2014-05-26            4648.01                   0   
14602      2012-08-27   2015-08-28             631.69                 181   
14603      2012-02-08   2015-02-09             190.39                 179   
14604      2012-08-30   2015-08-31              19.34                   0   
14605      2009-12-18   2015-12-21             762.41                   0   

       forecast_discount_energy  forecast_meter_rent_12m  \
0                           0.0                     1.78   
1                           0.0                    16.27   
2                           0.0                    38.72   
3                           0.0                    19.83   
4                           0.0                   131.73   
...                         ...                      ...   
14601                       0.0                    18.57   
14602                       0.0                   144.03   
14603                       0.0                   129.60   
14604                       0.0                     7.18   
14605                       0.0                     1.07   

       forecast_price_energy_off_peak  forecast_price_energy_peak  \
0                            0.114481                    0.098142   
1                            0.145711                    0.000000   
2                            0.165794                    0.087899   
3                            0.146694                    0.000000   
4                            0.116900                    0.100015   
...                               ...                         ...   
14601                        0.138305                    0.000000   
14602                    

In [3]:
price_df=pd.read_csv('price_data.csv')
price_df

id  price_date  price_off_peak_var  \
0       038af19179925da21a25619c5a24b745  2015-01-01            0.151367   
1       038af19179925da21a25619c5a24b745  2015-02-01            0.151367   
2       038af19179925da21a25619c5a24b745  2015-03-01            0.151367   
3       038af19179925da21a25619c5a24b745  2015-04-01            0.149626   
4       038af19179925da21a25619c5a24b745  2015-05-01            0.149626   
...                                  ...         ...                 ...   
192997  16f51cdc2baa19af0b940ee1b3dd17d5  2015-08-01            0.119916   
192998  16f51cdc2baa19af0b940ee1b3dd17d5  2015-09-01            0.119916   
192999  16f51cdc2baa19af0b940ee1b3dd17d5  2015-10-01            0.119916   
193000  16f51cdc2baa19af0b940ee1b3dd17d5  2015-11-01            0.119916   
193001  16f51cdc2baa19af0b940ee1b3dd17d5  2015-12-01            0.119916   

        price_peak_var  price_mid_peak_var  price_off_peak_fix  \
0             0.000000            0.000000           44.266931   
1             0.000000            0.000000           44.266931   
2             0.000000            0.000000           44.266931   
3             0.000000            0.000000           44.266931   
4             0.000000            0.000000           44.266931   
...                ...                 ...                 ...   
192997        0.102232            0.076257           40.728885   
192998        0.102232            0.076257           40.728885   
192999        0.102232            0.076257           40.728885   
193000        0.102232            0.076257           40.728885   
193001        0.102232            0.076257           40.728885   

        price_peak_fix  price_mid_peak_fix  
0              0.00000            0.000000  
1              0.00000            0.000000  
2              0.00000            0.000000  
3              0.00000            0.000000  
4              0.00000            0.000000  
...                ...                 ...  
192997        24.43733           16.291555  
192998        24.43733           16.291555  
192999        24.43733           16.291555  
193000        24.43733           16.291555  
193001        24.43733           16.291555  

[193002 rows x 8 columns]

In [4]:
#Adding prices to avoid duplicated rows for the same customer.
sum_df=price_df.groupby('id')['price_off_peak_var', 'price_peak_var','price_mid_peak_var', 'price_off_peak_fix',
                   'price_peak_fix','price_mid_peak_fix'].sum()
sum_df

C:\myanaconda\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  


price_off_peak_var  price_peak_var  \
id                                                                     
0002203ffbb812588b632b9e628cc38d            1.492061        1.245525   
0004351ebdd665e6ee664792efc4fd13            1.757118        0.000000   
0010bcc39e42b3c2131ed2ce55246e3c            2.178702        0.000000   
0010ee3855fdea87602a5b7aba8e42de            1.425085        1.179509   
00114d74e963e47177db89bc70108537            1.775110        0.000000   
...                                              ...             ...   
ffef185810e44254c3a4c6395e6b4d8a            1.666360        1.381496   
fffac626da707b1b5ab11e8431a4d0a2            1.765642        0.000000   
fffc0cacd305dd51f316424bbb08d1bd            1.846544        1.553959   
fffe4f5646aa39c7f97f95ae2679ce64            1.486297        1.241987   
ffff7fa066f1fb305ae285bb03bf325a            1.504321        1.258739   

                                  price_mid_peak_var  price_off_peak_fix  \
id                                                                         
0002203ffbb812588b632b9e628cc38d            0.877924          488.420789   
0004351ebdd665e6ee664792efc4fd13            0.000000          532.625404   
0010bcc39e42b3c2131ed2ce55246e3c            0.000000          543.836520   
0010ee3855fdea87602a5b7aba8e42de            0.828384          487.769130   
00114d74e963e47177db89bc70108537            0.000000          531.203164   
...                                              ...                 ...   
ffef185810e44254c3a4c6395e6b4d8a            0.969363          490.757129   
fffac626da707b1b5ab11e8431a4d0a2            0.000000          531.736506   
fffc0cacd305dd51f316424bbb08d1bd            1.138100          493.922047   
fffe4f5646aa39c7f97f95ae2679ce64            0.884823          487.280394   
ffff7fa066f1fb305ae285bb03bf325a            0.907620          487.769130   

                                  price_peak_fix  price_mid_peak_fix  
id                                                                    
0002203ffbb812588b632b9e628cc38d      293.052462          195.368327  
0004351ebdd665e6ee664792efc4fd13        0.000000            0.000000  
0010bcc39e42b3c2131ed2ce55246e3c        0.000000            0.000000  
0010ee3855fdea87602a5b7aba8e42de      292.661462          195.107656  
00114d74e963e47177db89bc70108537        0.000000            0.000000  
...                                          ...                 ...  
ffef185810e44254c3a4c6395e6b4d8a      295.649472          198.095667  
fffac626da707b1b5ab11e8431a4d0a2        0.000000            0.000000  
fffc0cacd305dd51f316424bbb08d1bd      298.749219          201.162833  
fffe4f5646aa39c7f97f95ae2679ce64      292.368206          194.912140  
ffff7fa066f1fb305ae285bb03bf325a      292.661462          195.107656  

[16096 rows x 6 columns]

- We figured out that for each client there are multiple rows for the price data so, we added them to avoid duplication and confusion.

In [5]:
#Joining two tables into a dataframe that we can start working on.
df=pd.merge(client_df,sum_df,on='id')
df

id                     channel_sales  \
0      24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1      d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2      764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3      bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4      149d57cf92fc41cf94415803a877cb4b                           MISSING   
...                                 ...                               ...   
14601  18463073fb097fc0ac5d3e040f356987  foosdfpfkusacimwkcsosbicdxkicaua   
14602  d0a6f71671571ed83b2645d23af6de00  foosdfpfkusacimwkcsosbicdxkicaua   
14603  10e6828ddd62cbcf687cb74928c4c2d2  foosdfpfkusacimwkcsosbicdxkicaua   
14604  1cf20fd6206d7678d5bcafd28c53b4db  foosdfpfkusacimwkcsosbicdxkicaua   
14605  563dde550fd624d7352f3de77c0cdfcd                           MISSING   

       cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0             0         54946                0  2013-06-15  2016-06-15   
1          4660             0                0  2009-08-21  2016-08-30   
2           544             0                0  2010-04-16  2016-04-16   
3          1584             0                0  2010-03-30  2016-03-30   
4          4425             0              526  2010-01-13  2016-03-07   
...         ...           ...              ...         ...         ...   
14601     32270         47940                0  2012-05-24  2016-05-08   
14602      7223             0              181  2012-08-27  2016-08-27   
14603      1844             0              179  2012-02-08  2016-02-07   
14604       131             0                0  2012-08-30  2016-08-30   
14605      8730             0                0  2009-12-18  2016-12-17   

      date_modif_prod date_renewal  forecast_cons_12m  forecast_cons_year  \
0          2015-11-01   2015-06-23               0.00                   0   
1          2009-08-21   2015-08-31             189.95                   0   
2          2010-04-16   2015-04-17              47.96                   0   
3          2010-03-30   2015-03-31             240.04                   0   
4          2010-01-13   2015-03-09             445.75                 526   
...               ...          ...                ...                 ...   
14601      2015-05-08   2014-05-26            4648.01                   0   
14602      2012-08-27   2015-08-28             631.69                 181   
14603      2012-02-08   2015-02-09             190.39                 179   
14604      2012-08-30   2015-08-31              19.34                   0   
14605      2009-12-18   2015-12-21             762.41                   0   

       forecast_discount_energy  forecast_meter_rent_12m  \
0                           0.0                     1.78   
1                           0.0                    16.27   
2                           0.0                    38.72   
3                           0.0                    19.83   
4                           0.0                   131.73   
...                         ...                      ...   
14601                       0.0                    18.57   
14602                       0.0                   144.03   
14603                       0.0                   129.60   
14604                       0.0                     7.18   
14605                       0.0                     1.07   

       forecast_price_energy_off_peak  forecast_price_energy_peak  \
0                            0.114481                    0.098142   
1                            0.145711                    0.000000   
2                            0.165794                    0.087899   
3                            0.146694                    0.000000   
4                            0.116900                    0.100015   
...                               ...                         ...   
14601                        0.138305                    0.000000   
14602                    

In [6]:
#Checking by 'id' that client has only one row.
df[df['id']=='24011ae4ebbe3035111d65fa7c15bc57']

id                     channel_sales  \
0  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   

   cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0         0         54946                0  2013-06-15  2016-06-15   

  date_modif_prod date_renewal  forecast_cons_12m  forecast_cons_year  \
0      2015-11-01   2015-06-23                0.0                   0   

   forecast_discount_energy  forecast_meter_rent_12m  \
0                       0.0                     1.78   

   forecast_price_energy_off_peak  forecast_price_energy_peak  \
0                        0.114481                    0.098142   

   forecast_price_pow_off_peak has_gas  imp_cons  margin_gross_pow_ele  \
0                    40.606701       t       0.0                 25.44   

   margin_net_pow_ele  nb_prod_act  net_margin  num_years_antig  \
0               25.44            2      678.99                3   

                          origin_up  pow_max  churn  price_off_peak_var  \
0  lxidpiddsbxsbosboudacockeimpuepw   43.648      1            1.497445   

   price_peak_var  price_mid_peak_var  price_off_peak_fix  price_peak_fix  \
0        1.208992            0.798365          491.307186      268.224124   

   price_mid_peak_fix  
0          178.816084

In [7]:
#Number of rows and columns(features)
df.shape


(14606, 32)

In [8]:
#checking features of all the data available
df.describe().astype('int64') 

cons_12m  cons_gas_12m  cons_last_month  forecast_cons_12m  \
count     14606         14606            14606              14606   
mean     159220         28092            16090               1868   
std      573465        162973            64364               2387   
min           0             0                0                  0   
25%        5674             0                0                494   
50%       14115             0              792               1112   
75%       40763             0             3383               2401   
max     6207104       4154590           771203              82902   

       forecast_cons_year  forecast_discount_energy  forecast_meter_rent_12m  \
count               14606                     14606                    14606   
mean                 1399                         0                       63   
std                  3247                         5                       66   
min                     0                         0                        0   
25%                     0                         0                       16   
50%                   314                         0                       18   
75%                  1745                         0                      131   
max                175375                        30                      599   

       forecast_price_energy_off_peak  forecast_price_energy_peak  \
count                           14606                       14606   
mean                                0                           0   
std                                 0                           0   
min                                 0                           0   
25%                                 0                           0   
50%                                 0                           0   
75%                                 0                           0   
max                                 0                           0   

       forecast_price_pow_off_peak  imp_cons  margin_gross_pow_ele  \
count                        14606     14606                 14606   
mean                            43       152                    24   
std                              4       341                    20   
min                              0         0                     0   
25%                             40         0                    14   
50%                             44        37                    21   
75%                             44       193                    29   
max                             59     15042                   374   

       margin_net_pow_ele  nb_prod_act  net_margin  num_years_antig  pow_max  \
count               14606        14606       14606            14606    14606   
mean                   24            1         189                4       18   
std                    20            0         311                1       13   
min                     0            1           0                1        3   
25%                    14            1          50                4       12   
50%                    21            1         112                5       13   
75%                    29            1         243                6       19   
max                   374           32       24570               13      320   

       churn  price_off_peak_var  price_peak_var  price_mid_peak_var  \
count  14606               14606           14606               14606   
mean       0                   1               0                   0   
std        0                   0               0                   0   
min        0                   0               0                   0   
25%        0                   1               0                   0   
50%        0                   1               1                   0   
75%        0                   1               1                   0   
max        1                   3               2                   1   

       price_off_peak_fix  price_peak_fix

In [9]:
#checking duplicated rows
df.duplicated().sum()

0

In [10]:
#check all the data types and check if there is any null data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14606 entries, 0 to 14605
Data columns (total 32 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              14606 non-null  object 
 1   channel_sales                   14606 non-null  object 
 2   cons_12m                        14606 non-null  int64  
 3   cons_gas_12m                    14606 non-null  int64  
 4   cons_last_month                 14606 non-null  int64  
 5   date_activ                      14606 non-null  object 
 6   date_end                        14606 non-null  object 
 7   date_modif_prod                 14606 non-null  object 
 8   date_renewal                    14606 non-null  object 
 9   forecast_cons_12m               14606 non-null  float64
 10  forecast_cons_year              14606 non-null  int64  
 11  forecast_discount_energy        14606 non-null  float64
 12  forecast_meter_rent_12m         

In [11]:
print(df.isnull().sum().sum())

0


In [12]:
#checking if there is multiple rows for the same customer
df['id'].duplicated().sum()

0

## Conclusion of data wrangling:
After preparing the two tables and merging them into one dataframe that we can work on:
- There are 14606 rows and 32 column.
- There are 0 null values.
- There are 0 customers that have multiple rows with different data.
- There are no duplicated rows.

### Part II: Data Cleaning

In [13]:
df.keys()


Index(['id', 'channel_sales', 'cons_12m', 'cons_gas_12m', 'cons_last_month',
       'date_activ', 'date_end', 'date_modif_prod', 'date_renewal',
       'forecast_cons_12m', 'forecast_cons_year', 'forecast_discount_energy',
       'forecast_meter_rent_12m', 'forecast_price_energy_off_peak',
       'forecast_price_energy_peak', 'forecast_price_pow_off_peak', 'has_gas',
       'imp_cons', 'margin_gross_pow_ele', 'margin_net_pow_ele', 'nb_prod_act',
       'net_margin', 'num_years_antig', 'origin_up', 'pow_max', 'churn',
       'price_off_peak_var', 'price_peak_var', 'price_mid_peak_var',
       'price_off_peak_fix', 'price_peak_fix', 'price_mid_peak_fix'],
      dtype='object')

In [14]:
#remove the unnecessary columns
df.drop(['margin_net_pow_ele','date_renewal','date_modif_prod','nb_prod_act', 'pow_max'],axis=1)

id                     channel_sales  \
0      24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1      d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2      764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3      bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4      149d57cf92fc41cf94415803a877cb4b                           MISSING   
...                                 ...                               ...   
14601  18463073fb097fc0ac5d3e040f356987  foosdfpfkusacimwkcsosbicdxkicaua   
14602  d0a6f71671571ed83b2645d23af6de00  foosdfpfkusacimwkcsosbicdxkicaua   
14603  10e6828ddd62cbcf687cb74928c4c2d2  foosdfpfkusacimwkcsosbicdxkicaua   
14604  1cf20fd6206d7678d5bcafd28c53b4db  foosdfpfkusacimwkcsosbicdxkicaua   
14605  563dde550fd624d7352f3de77c0cdfcd                           MISSING   

       cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0             0         54946                0  2013-06-15  2016-06-15   
1          4660             0                0  2009-08-21  2016-08-30   
2           544             0                0  2010-04-16  2016-04-16   
3          1584             0                0  2010-03-30  2016-03-30   
4          4425             0              526  2010-01-13  2016-03-07   
...         ...           ...              ...         ...         ...   
14601     32270         47940                0  2012-05-24  2016-05-08   
14602      7223             0              181  2012-08-27  2016-08-27   
14603      1844             0              179  2012-02-08  2016-02-07   
14604       131             0                0  2012-08-30  2016-08-30   
14605      8730             0                0  2009-12-18  2016-12-17   

       forecast_cons_12m  forecast_cons_year  forecast_discount_energy  \
0                   0.00                   0                       0.0   
1                 189.95                   0                       0.0   
2                  47.96                   0                       0.0   
3                 240.04                   0                       0.0   
4                 445.75                 526                       0.0   
...                  ...                 ...                       ...   
14601            4648.01                   0                       0.0   
14602             631.69                 181                       0.0   
14603             190.39                 179                       0.0   
14604              19.34                   0                       0.0   
14605             762.41                   0                       0.0   

       forecast_meter_rent_12m  forecast_price_energy_off_peak  \
0                         1.78                        0.114481   
1                        16.27                        0.145711   
2                        38.72                        0.165794   
3                        19.83                        0.146694   
4                       131.73                        0.116900   
...                        ...                             ...   
14601                    18.57                        0.138305   
14602                   144.03                        0.100167   
14603                   129.60                        0.116900   
14604                     7.18                        0.145711   
14605                     1.07                        0.167086   

       forecast_price_energy_peak  forecast_price_pow_off_peak has_gas  \
0                        0.098142                    40.606701       t   
1                        0.000000                    44.311378       f   
2                        0.087899                    44.311378       f   
3                        0.000000                    44.311378       f   
4                        0.100015                    40.606701       f   
...                           ...                          ...     ...   
14601                  

In [15]:
#Changing all date columns into datetime format
df["date_activ"] = pd.to_datetime(df["date_activ"], format='%Y-%m-%d')
df["date_end"] = pd.to_datetime(df["date_end"], format='%Y-%m-%d')
price_df['price_date'] = pd.to_datetime(price_df['price_date'], format='%Y-%m-%d')

In [16]:
#Modifying values into boolean type in the 'has_gas' column
df['has_gas'].unique()

array(['t', 'f'], dtype=object)

In [17]:
def has_gas (row):
    if(row['has_gas']=='t'):
        return True
    else:
        return False
df['has_gas']=df.apply(has_gas,axis=1)
df

id                     channel_sales  \
0      24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1      d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2      764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3      bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4      149d57cf92fc41cf94415803a877cb4b                           MISSING   
...                                 ...                               ...   
14601  18463073fb097fc0ac5d3e040f356987  foosdfpfkusacimwkcsosbicdxkicaua   
14602  d0a6f71671571ed83b2645d23af6de00  foosdfpfkusacimwkcsosbicdxkicaua   
14603  10e6828ddd62cbcf687cb74928c4c2d2  foosdfpfkusacimwkcsosbicdxkicaua   
14604  1cf20fd6206d7678d5bcafd28c53b4db  foosdfpfkusacimwkcsosbicdxkicaua   
14605  563dde550fd624d7352f3de77c0cdfcd                           MISSING   

       cons_12m  cons_gas_12m  cons_last_month date_activ   date_end  \
0             0         54946                0 2013-06-15 2016-06-15   
1          4660             0                0 2009-08-21 2016-08-30   
2           544             0                0 2010-04-16 2016-04-16   
3          1584             0                0 2010-03-30 2016-03-30   
4          4425             0              526 2010-01-13 2016-03-07   
...         ...           ...              ...        ...        ...   
14601     32270         47940                0 2012-05-24 2016-05-08   
14602      7223             0              181 2012-08-27 2016-08-27   
14603      1844             0              179 2012-02-08 2016-02-07   
14604       131             0                0 2012-08-30 2016-08-30   
14605      8730             0                0 2009-12-18 2016-12-17   

      date_modif_prod date_renewal  forecast_cons_12m  forecast_cons_year  \
0          2015-11-01   2015-06-23               0.00                   0   
1          2009-08-21   2015-08-31             189.95                   0   
2          2010-04-16   2015-04-17              47.96                   0   
3          2010-03-30   2015-03-31             240.04                   0   
4          2010-01-13   2015-03-09             445.75                 526   
...               ...          ...                ...                 ...   
14601      2015-05-08   2014-05-26            4648.01                   0   
14602      2012-08-27   2015-08-28             631.69                 181   
14603      2012-02-08   2015-02-09             190.39                 179   
14604      2012-08-30   2015-08-31              19.34                   0   
14605      2009-12-18   2015-12-21             762.41                   0   

       forecast_discount_energy  forecast_meter_rent_12m  \
0                           0.0                     1.78   
1                           0.0                    16.27   
2                           0.0                    38.72   
3                           0.0                    19.83   
4                           0.0                   131.73   
...                         ...                      ...   
14601                       0.0                    18.57   
14602                       0.0                   144.03   
14603                       0.0                   129.60   
14604                       0.0                     7.18   
14605                       0.0                     1.07   

       forecast_price_energy_off_peak  forecast_price_energy_peak  \
0                            0.114481                    0.098142   
1                            0.145711                    0.000000   
2                            0.165794                    0.087899   
3                            0.146694                    0.000000   
4                            0.116900                    0.100015   
...                               ...                         ...   
14601                        0.138305                    0.000000   
14602                        0.100167            

In [18]:
#Adding column that determines whether client has only electricity contract type
def cons_12m (row):
    if(row['cons_12m']==0):
        return False
    else:
        return True
df['has_electricity']=df.apply(cons_12m,axis=1)
df

id                     channel_sales  \
0      24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1      d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2      764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3      bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4      149d57cf92fc41cf94415803a877cb4b                           MISSING   
...                                 ...                               ...   
14601  18463073fb097fc0ac5d3e040f356987  foosdfpfkusacimwkcsosbicdxkicaua   
14602  d0a6f71671571ed83b2645d23af6de00  foosdfpfkusacimwkcsosbicdxkicaua   
14603  10e6828ddd62cbcf687cb74928c4c2d2  foosdfpfkusacimwkcsosbicdxkicaua   
14604  1cf20fd6206d7678d5bcafd28c53b4db  foosdfpfkusacimwkcsosbicdxkicaua   
14605  563dde550fd624d7352f3de77c0cdfcd                           MISSING   

       cons_12m  cons_gas_12m  cons_last_month date_activ   date_end  \
0             0         54946                0 2013-06-15 2016-06-15   
1          4660             0                0 2009-08-21 2016-08-30   
2           544             0                0 2010-04-16 2016-04-16   
3          1584             0                0 2010-03-30 2016-03-30   
4          4425             0              526 2010-01-13 2016-03-07   
...         ...           ...              ...        ...        ...   
14601     32270         47940                0 2012-05-24 2016-05-08   
14602      7223             0              181 2012-08-27 2016-08-27   
14603      1844             0              179 2012-02-08 2016-02-07   
14604       131             0                0 2012-08-30 2016-08-30   
14605      8730             0                0 2009-12-18 2016-12-17   

      date_modif_prod date_renewal  forecast_cons_12m  forecast_cons_year  \
0          2015-11-01   2015-06-23               0.00                   0   
1          2009-08-21   2015-08-31             189.95                   0   
2          2010-04-16   2015-04-17              47.96                   0   
3          2010-03-30   2015-03-31             240.04                   0   
4          2010-01-13   2015-03-09             445.75                 526   
...               ...          ...                ...                 ...   
14601      2015-05-08   2014-05-26            4648.01                   0   
14602      2012-08-27   2015-08-28             631.69                 181   
14603      2012-02-08   2015-02-09             190.39                 179   
14604      2012-08-30   2015-08-31              19.34                   0   
14605      2009-12-18   2015-12-21             762.41                   0   

       forecast_discount_energy  forecast_meter_rent_12m  \
0                           0.0                     1.78   
1                           0.0                    16.27   
2                           0.0                    38.72   
3                           0.0                    19.83   
4                           0.0                   131.73   
...                         ...                      ...   
14601                       0.0                    18.57   
14602                       0.0                   144.03   
14603                       0.0                   129.60   
14604                       0.0                     7.18   
14605                       0.0                     1.07   

       forecast_price_energy_off_peak  forecast_price_energy_peak  \
0                            0.114481                    0.098142   
1                            0.145711                    0.000000   
2                            0.165794                    0.087899   
3                            0.146694                    0.000000   
4                            0.116900                    0.100015   
...                               ...                         ...   
14601                        0.138305                    0.000000   
14602                        0.100167            

In [19]:
#Adding column that determines whether client has both electricity and gas contract type
def has_both (row):
    if(row['has_electricity']==True and row['has_gas']==True):
        return True
    else:
        return False
df['has_both']=df.apply(has_both,axis=1)
df

id                     channel_sales  \
0      24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1      d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2      764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3      bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4      149d57cf92fc41cf94415803a877cb4b                           MISSING   
...                                 ...                               ...   
14601  18463073fb097fc0ac5d3e040f356987  foosdfpfkusacimwkcsosbicdxkicaua   
14602  d0a6f71671571ed83b2645d23af6de00  foosdfpfkusacimwkcsosbicdxkicaua   
14603  10e6828ddd62cbcf687cb74928c4c2d2  foosdfpfkusacimwkcsosbicdxkicaua   
14604  1cf20fd6206d7678d5bcafd28c53b4db  foosdfpfkusacimwkcsosbicdxkicaua   
14605  563dde550fd624d7352f3de77c0cdfcd                           MISSING   

       cons_12m  cons_gas_12m  cons_last_month date_activ   date_end  \
0             0         54946                0 2013-06-15 2016-06-15   
1          4660             0                0 2009-08-21 2016-08-30   
2           544             0                0 2010-04-16 2016-04-16   
3          1584             0                0 2010-03-30 2016-03-30   
4          4425             0              526 2010-01-13 2016-03-07   
...         ...           ...              ...        ...        ...   
14601     32270         47940                0 2012-05-24 2016-05-08   
14602      7223             0              181 2012-08-27 2016-08-27   
14603      1844             0              179 2012-02-08 2016-02-07   
14604       131             0                0 2012-08-30 2016-08-30   
14605      8730             0                0 2009-12-18 2016-12-17   

      date_modif_prod date_renewal  forecast_cons_12m  forecast_cons_year  \
0          2015-11-01   2015-06-23               0.00                   0   
1          2009-08-21   2015-08-31             189.95                   0   
2          2010-04-16   2015-04-17              47.96                   0   
3          2010-03-30   2015-03-31             240.04                   0   
4          2010-01-13   2015-03-09             445.75                 526   
...               ...          ...                ...                 ...   
14601      2015-05-08   2014-05-26            4648.01                   0   
14602      2012-08-27   2015-08-28             631.69                 181   
14603      2012-02-08   2015-02-09             190.39                 179   
14604      2012-08-30   2015-08-31              19.34                   0   
14605      2009-12-18   2015-12-21             762.41                   0   

       forecast_discount_energy  forecast_meter_rent_12m  \
0                           0.0                     1.78   
1                           0.0                    16.27   
2                           0.0                    38.72   
3                           0.0                    19.83   
4                           0.0                   131.73   
...                         ...                      ...   
14601                       0.0                    18.57   
14602                       0.0                   144.03   
14603                       0.0                   129.60   
14604                       0.0                     7.18   
14605                       0.0                     1.07   

       forecast_price_energy_off_peak  forecast_price_energy_peak  \
0                            0.114481                    0.098142   
1                            0.145711                    0.000000   
2                            0.165794                    0.087899   
3                            0.146694                    0.000000   
4                            0.116900                    0.100015   
...                               ...                         ...   
14601                        0.138305                    0.000000   
14602                        0.100167            

In [20]:
#saving updated dataset for answering questions using SQL
df.to_csv('merged_data.csv' , index=False)
price_df.to_csv('modified_price_data.csv', index=False)